### PR TITLE
[Impeller] Discard invalid command buffer handles after destroying a command pool

### PR DESCRIPTION
**Pull request by jason-simmons**
Fixes https://github.com/flutter/flutter/issues/131522

### CODE REVIEW

**Review by gaaclarke — COMMENTED** [review]
LGTM, any idea how we could test this?

File: impeller/renderer/backend/vulkan/command_pool_vk.cc [hunk 1]
```diff
@@ -92,8 +92,15 @@ bool CommandPoolVK::IsValid() const {
 
 void CommandPoolVK::Reset() {
   Lock lock(buffers_to_collect_mutex_);
-  GarbageCollectBuffersIfAble();
   graphics_pool_.reset();
+
+  // When the command pool is destroyed, all of its command buffers are freed.
+  // Handles allocated from that pool are now invalid and must be discarded.
+  for (vk::UniqueCommandBuffer& buffer : buffers_to_collect_) {
+    buffer.release();
```
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/command_pool_vk.cc:100]: But the handle to the buffer will still be leaked right? Unless this is a reset.

Comment (jason-simmons) [reply impeller/renderer/backend/vulkan/command_pool_vk.cc:100]: Earlier this function calls `graphics_pool_.reset()`, which destroys the Vulkan command pool and frees all of the pool's buffers.

After that the `buffer` handle is invalid, and the Vulkan validation layers will warn about a call to `vkFreeCommandBuffers` using the handle.

**Review by chinmaygarde — APPROVED** [review]

**Comment by zanderso** [comment]
@jason-simmons @chinmaygarde I see that @gaaclarke had a question about testing on this PR that went unanswered, so I'll go ahead and echo that. IIRC we there's a test harness for Vulkan we can use for stuff like this, right?

**Comment by jason-simmons** [comment]
@gaaclarke had created a framework for mocking a `ContextVK` (https://github.com/flutter/engine/blob/main/impeller/renderer/backend/vulkan/test/mock_vulkan.h).  But I don't know if we have any harness that can invoke Vulkan APIs and check for validation layer violations.

**Comment by chinmaygarde** [comment]
My bad. We were going through this in triage and missed the call. I am not sure how to test this though TBH. We'd need to mock out Vulkan entrypoints.

**Comment by gaaclarke** [comment]
I think `mock_vulkan.h` should be usable.  Every single call to vulkan can be captured, so you could check the calls to `vkFreeCommandBuffers`.  It would require a bit of plumbing but hopefully if more people used it, it will get easier to use.  Unfortunately I think I did the plumbing at some point, but then dropped the PR so I never landed it.

Another option would be to hook into `DebugReportVK` and actually grab the validation layers violation.  We are using validation layers to run impeller_unittests if I remember correctly.  That hasn't been plumbed up, but once someone does it it will be easier for the next person.

**Comment by gaaclarke** [comment]
> We'd need to mock out Vulkan entrypoints.

That's what `mock_vulkan.h` does.

**Pull request merged**